### ISSUE TITLE
Bias between Save_weights and Load_weights

### ISSUE THREAD

**Issue by yyaodong** [fchollet/keras#655]
Hi All,

I want to save the model which shows the best result on validation set during model.fit() process. I create a callback that uses model.save_weights to save the best model to a hdf5 file. However, after the whole training is finished, by re-constructing the best model via model.load_weights, I cannot repeat that best result on the validation set anymore. The reconstructed model does not produce the same performance that it shows in the fitting process. The bias exists in model.set_weights and model.get_weights as well.

My current solution is to use the copy.deepcopy function to copy the whole "model" when a model with better performance is found. But it largely deteriorates the speed. I was wondering is this a bug or I miss some basic things of the "model" during re-construction?

Thanks you very much.
Yang

**Comment by fchollet** [comment]
What is your model architecture?

**Comment by yyaodong** [comment]
My model is very straightforward:

    model = Sequential()
    model.add(Dense(X_valid.shape[1], hiddenLayers[0],init='glorot_uniform')) # matrix + bias
    model.add(PReLU((hiddenLayers[0],)))
    model.add(BatchNormalization((hiddenLayers[0],)))
    model.add(Dropout(0.5))
    model.add(Dense(hiddenLayers[-1], nb_classes))
    model.add(Activation('softmax'))
    model.compile(loss='categorical_crossentropy', optimizer=Adam()) 

    callback = keras.callbacks.ModelCheckpoint(filepath, verbose=0, save_best_only=False)

    model.fit(X_train, Ytrain, batch_size=16, nb_epoch=10, verbose=2, validation_data=(X_valid, Yvalid), callbacks=[callback], show_accuracy=True)

**Comment by fchollet** [comment]
The issue here would be the `BatchNormalization` layer, as it contains parameters that are not saved / loaded with `save_weights` and `load_weights`. We will release a fix.

**Comment by fchollet** [comment]
This is now fixed.
